# Building a Logistic Regression

Create a logistic regression based on the bank data provided. 

The data is based on the marketing campaign efforts of a Portuguese banking institution. The classification goal is to predict if the client will subscribe a term deposit (variable y).

Note that the first column of the dataset is the index.

Source: [Moro et al., 2014] S. Moro, P. Cortez and P. Rita. A Data-Driven Approach to Predict the Success of Bank Telemarketing. Decision Support Systems, Elsevier, 62:22-31, June 2014


## Import the relevant libraries

In [3]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sn
sn.set()

## Load the data

Load the ‘Example_bank_data.csv’ dataset.

In [4]:
data = pd.read_csv('C:/Users/---/Desktop/Projects/Practice/365_DataScience/Bank_data.csv')
data

Unnamed: 0  interest_rate  credit  march  may  previous  duration    y
0             0          1.334     0.0    1.0  0.0       0.0     117.0   no
1             1          0.767     0.0    0.0  2.0       1.0     274.0  yes
2             2          4.858     0.0    1.0  0.0       0.0     167.0   no
3             3          4.120     0.0    0.0  0.0       0.0     686.0  yes
4             4          4.856     0.0    1.0  0.0       0.0     157.0   no
..          ...            ...     ...    ...  ...       ...       ...  ...
513         513          1.334     0.0    1.0  0.0       0.0     204.0   no
514         514          0.861     0.0    0.0  2.0       1.0     806.0  yes
515         515          0.879     0.0    0.0  0.0       0.0     290.0   no
516         516          0.877     0.0    0.0  5.0       1.0     473.0  yes
517         517          4.965     0.0    0.0  0.0       0.0     142.0   no

[518 rows x 8 columns]

We want to know whether the bank marketing strategy was successful, so we need to transform the outcome variable into 0s and 1s in order to perform a logistic regression.

In [5]:
data['y'] = data['y'].map({'yes':1,'no':0})

In [6]:
data

Unnamed: 0  interest_rate  credit  march  may  previous  duration  y
0             0          1.334     0.0    1.0  0.0       0.0     117.0  0
1             1          0.767     0.0    0.0  2.0       1.0     274.0  1
2             2          4.858     0.0    1.0  0.0       0.0     167.0  0
3             3          4.120     0.0    0.0  0.0       0.0     686.0  1
4             4          4.856     0.0    1.0  0.0       0.0     157.0  0
..          ...            ...     ...    ...  ...       ...       ... ..
513         513          1.334     0.0    1.0  0.0       0.0     204.0  0
514         514          0.861     0.0    0.0  2.0       1.0     806.0  1
515         515          0.879     0.0    0.0  0.0       0.0     290.0  0
516         516          0.877     0.0    0.0  5.0       1.0     473.0  1
517         517          4.965     0.0    0.0  0.0       0.0     142.0  0

[518 rows x 8 columns]

### Declare the dependent and independent variables

In [7]:
y = data['y']
x1 = data['duration']

### Simple Logistic Regression

Run the regression and visualize it on a scatter plot (no need to plot the line).

In [8]:
x = sm.add_constant(x1)
reg_log = sm.Logit(y,x)
result_log = reg_log.fit()

Optimization terminated successfully.
         Current function value: 0.546118
         Iterations 7


C:\Users\---\Anaconda3\lib\site-packages\numpy\core\fromnumeric.py:2389: FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
  return ptp(axis=axis, out=out, **kwargs)


In [9]:
# so we managed to fit the regression
# it took 7 iterations and 0.54 current function value
# FYI in statsmodel the maximum iterations is 35 
# so if the iterations is higher than that its will get a warning message
result_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                  518
Model:                          Logit   Df Residuals:                      516
Method:                           MLE   Df Model:                            1
Date:                Thu, 15 Oct 2020   Pseudo R-squ.:                  0.2121
Time:                        14:56:33   Log-Likelihood:                -282.89
converged:                       True   LL-Null:                       -359.05
Covariance Type:            nonrobust   LLR p-value:                 5.387e-35
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.7001      0.192     -8.863      0.000      -2.076      -1.324
duration       0.0051      0.001      9.159      0.000       0.004       0.006
==============================================================================
"""

 as we can see in the statsmodel summary its almost similar to the OLS one 
 sometimes when the summary on statsmodel does not working try to import this one
 #### from scipy import stats
 #### stats.chisqprob = lambda chisq, df: stats.chi2.sf(chisq,df)

## Lets do the accuracy test

In [10]:
result_log.predict()

array([0.24936904, 0.42571348, 0.30019503, 0.85898342, 0.28956627,
       0.2580803 , 0.21914028, 0.16614452, 0.86976608, 0.31987434,
       0.55276776, 0.22532533, 0.30234734, 0.39480294, 0.35411888,
       0.489064  , 0.29167414, 0.25225084, 0.50567672, 0.82802726,
       0.84413091, 0.5890681 , 0.45848474, 0.40215497, 0.44961416,
       0.20546539, 0.4157464 , 0.42321571, 0.58286683, 0.2580803 ,
       0.17928751, 0.2580803 , 0.85711543, 0.43322907, 0.293791  ,
       0.32658401, 0.98373368, 0.21652721, 0.99665611, 0.62920727,
       0.2639969 , 0.49417474, 0.30885632, 0.27711089, 0.28746746,
       0.2474601 , 0.19643708, 0.30342677, 0.30559213, 0.4712014 ,
       0.87376626, 0.20132396, 0.22177596, 0.28642146, 0.25612742,
       0.49417474, 0.27916371, 0.20214723, 0.88872861, 0.53501481,
       0.23806331, 0.24556099, 0.90211616, 0.40092648, 0.51078751,
       0.45975425, 0.28019361, 0.30450837, 0.87432905, 0.23713725,
       0.18231618, 0.3706541 , 0.4724754 , 0.38144917, 0.54009

In [11]:
# lets format it to readable array
np.set_printoptions(formatter={'float':lambda x:"{0:0.2f}".format(x)})
result_log.predict()

# as we can see in the below array theres 0.00-1.00 value that is the probability

array([0.25, 0.43, 0.30, 0.86, 0.29, 0.26, 0.22, 0.17, 0.87, 0.32, 0.55,
       0.23, 0.30, 0.39, 0.35, 0.49, 0.29, 0.25, 0.51, 0.83, 0.84, 0.59,
       0.46, 0.40, 0.45, 0.21, 0.42, 0.42, 0.58, 0.26, 0.18, 0.26, 0.86,
       0.43, 0.29, 0.33, 0.98, 0.22, 1.00, 0.63, 0.26, 0.49, 0.31, 0.28,
       0.29, 0.25, 0.20, 0.30, 0.31, 0.47, 0.87, 0.20, 0.22, 0.29, 0.26,
       0.49, 0.28, 0.20, 0.89, 0.54, 0.24, 0.25, 0.90, 0.40, 0.51, 0.46,
       0.28, 0.30, 0.87, 0.24, 0.18, 0.37, 0.47, 0.38, 0.54, 0.26, 0.41,
       0.57, 0.29, 0.21, 0.34, 0.43, 0.30, 0.46, 0.41, 0.97, 0.30, 0.53,
       0.40, 0.23, 0.36, 0.22, 0.37, 0.32, 0.77, 0.81, 0.80, 0.20, 0.98,
       0.30, 0.93, 0.34, 0.89, 0.41, 0.50, 0.50, 0.42, 0.35, 0.42, 0.52,
       0.54, 0.25, 0.31, 0.21, 0.74, 0.18, 0.25, 0.88, 0.60, 0.48, 0.23,
       0.17, 0.40, 0.99, 0.48, 0.96, 0.36, 0.59, 0.34, 0.42, 0.29, 0.33,
       0.48, 0.90, 0.46, 0.68, 0.57, 0.43, 0.43, 0.31, 0.26, 0.23, 0.84,
       0.44, 0.84, 0.65, 0.38, 0.26, 1.00, 0.26, 0.

In [12]:
# there is a simple way to see the accuracy of the model by using a stats model prediction model
# we call this the confusion matrix, this matrix measure the error prediction of the model

result_log.pred_table ()

array([[204.00, 55.00],
       [104.00, 155.00]])

In [13]:
# lets make a confuison matrix table based on pred_table based on stats model

cm_df = pd.DataFrame(result_log.pred_table())
cm_df.columns = ['Predicted 0','Predicted 1']
cm_df = cm_df.rename(index ={0: 'Actual 0',1: 'Actual 1'})
cm_df

Predicted 0  Predicted 1
Actual 0        204.0         55.0
Actual 1        104.0        155.0

based on this confusion matrix we can interprete that 

204 + 155 = 359 (This is the result of model did its job well)

104 + 55 = 159 (This the result of model got confused or error)

so the ratio is 359:159

In [14]:
# if u want the percentage result this is the code

cm= np.array(cm_df)
accuracy_train = (cm[0,0]+cm[1,1])/cm.sum()
accuracy_train

0.693050193050193

so we get 69% accuracy

that is not a really good accuracy for the model so wen need to improve it by adding some more fucntion etc...